File: lua/stackless_coroutine/fibonacci_gen.lua
--[[
  fibonacci_gen.lua
  Stackless coroutine simulation (Lua coroutines are stackful by nature).

  Lua's coroutine.create/resume/yield are stackful — they preserve the entire
  call stack. This example wraps them to enforce a "stackless" discipline:
  yielding is only allowed at the top level of the generator, not from
  nested calls. This simulates the semantics of Python generators or
  JavaScript function* generators.

  Compare with: Python yield_fibonacci.py, JS generator_fibonacci.js
]]

local StacklessGenerator = {}
StacklessGenerator.__index = StacklessGenerator

function StacklessGenerator.new(fn)
  local self = setmetatable({
    co = coroutine.create(fn),
    done = false
  }, StacklessGenerator)
  return self
end

-- Get the next value from the generator (like Python's next())
function StacklessGenerator:next(...)
  if self.done then
    return nil
  end

  local ok, result = coroutine.resume(self.co, ...)
  if not ok then
    error("Generator error: " .. tostring(result))
  end

  if coroutine.status(self.co) == "dead" then
    self.done = true
  end

  return result
end

-- Fibonacci generator (stackless style: yield only at top level)
function fibonacci_stackless(limit)
  return StacklessGenerator.new(function()
    local a, b = 0, 1
    for i = 1, limit do
      coroutine.yield(a)  -- top-level yield only
      a, b = b, a + b
    end
  end)
end

-- WARNING: This will error because yield is inside a nested call.
-- This demonstrates the stackless limitation!
function fibonacci_bad_nested(limit)
  return StacklessGenerator.new(function()
    local function inner()
      coroutine.yield("oops")  -- This cannot work in true stackless!
    end
    inner()  -- Will error here if we try to yield from nested
  end)
end

-- Demo
print("=== Stackless Coroutine: Fibonacci Generator (Lua) ===\n")
print("(Simulating stackless semantics with Lua's stackful coroutines)\n")

print("--- fibonacci_stackless(20) ---")
local gen = fibonacci_stackless(20)
local i = 0
while true do
  local val = gen:next()
  if val == nil then break end
  print(string.format("  fib(%d) = %d", i, val))
  i = i + 1
end

print("\n--- Key insight ---")
print("Lua coroutines are NATIVELY STACKFUL: they can yield from")
print("any call depth. This demo enforces a 'top-level yield only'")
print("discipline to simulate stackless semantics.")
print()
print("Compare with Lua's native stackful fibonacci.lua which")
print("is more flexible (can yield from nested calls).")

File: scala/stackless_coroutine/FibonacciGen.scala
/**
 * FibonacciGen.scala
 *
 * Paradigm: Stackless Coroutine (via Iterator)
 *
 * Scala's Iterator is a stackless generator: state is held in the
 * heap-allocated iterator object, not on the call stack.
 * Compare with: Python yield, C++20 co_yield, JS generators
 *
 * NOTE: Uses Stream (Scala 2.11 compatible) instead of LazyList (2.13+).
 *
 * Compile:  scalac -d build stackless_coroutine/FibonacciGen.scala
 * Run:      scala -classpath build FibonacciGen
 */

object FibonacciGen {

  /** Fibonacci as a stackless Iterator — state on heap, not stack. */
  def fibonacciIterator(limit: Int): Iterator[Long] = new Iterator[Long] {
    private var a = 0L
    private var b = 1L
    private var count = 0

    override def hasNext: Boolean = count < limit

    override def next(): Long = {
      if (!hasNext) throw new NoSuchElementException()
      val result = a
      val next = a + b
      a = b
      b = next
      count += 1
      result
    }
  }

  /** Fibonacci via Stream (Scala 2.11 compatible). */
  def fibonacciStream: Stream[Long] = {
    def loop(a: Long, b: Long): Stream[Long] = a #:: loop(b, a + b)
    loop(0, 1)
  }

  def main(args: Array[String]): Unit = {
    println("=== Stackless Coroutine: Fibonacci Generator ===\n")

    println("--- fibonacciIterator(20) ---")
    val it = fibonacciIterator(20)
    var i = 0
    while (it.hasNext) {
      println("  fib(" + i + ") = " + it.next())
      i += 1
    }

    println("\n--- fibonacciStream (first 15) ---")
    fibonacciStream.take(15).zipWithIndex.foreach { case (v, idx) =>
      println("  fib(" + idx + ") = " + v)
    }

    println("\n=== Key insight ===")
    println("Iterator and Stream are stackless: each element's")
    println("state is heap-allocated. This is the same model as")
    println("Python generators and C++20 coroutines.")
  }
}
